lib/services/scoped_tokens.go

```diff
@@ -0,0 +1,115 @@
+	// CreateScopedToken creates a scoped join token.
+	CreateScopedToken(ctx context.Context, token *joiningv1.ScopedToken) (*joiningv1.ScopedToken, error)
+
+	// UpdateScopedToken updates a scoped join token.
+	UpdateScopedToken(ctx context.Context, token *joiningv1.ScopedToken) (*joiningv1.ScopedToken, error)
+
+	// GetScopedToken fetches a scoped join token by unique name
+	GetScopedToken(ctx context.Context, name string) (*joiningv1.ScopedToken, error)
+
+	// ListScopedTokens retrieves a paginated list of scoped join tokens
+	ListScopedTokens(ctx context.Context, pageSize int, pageToken string, filters *ScopedTokenFilters) ([]*joiningv1.ScopedToken, string, error)
+
+	// DeleteScopedToken deletes a named scoped join token. Imlementations must guarantee that
+	// this returns trace.NotFound error if the token doesn't exist
+	DeleteScopedToken(ctx context.Context, name string) error
+}
+
+// ValidateScopedToken checks that the given [joiningv1.ScopedToken] meets
+// the constraints required to be written to the backend.
+func ValidateScopedToken(token *joiningv1.ScopedToken) error {
+	if expected, actual := types.KindScopedToken, token.GetKind(); expected != actual {
+		return trace.BadParameter("expected kind %v, got %q", expected, actual)
+	}
+	if expected, actual := types.V1, token.GetVersion(); expected != actual {
+		return trace.BadParameter("expected version %v, got %q", expected, actual)
+	}
+	if expected, actual := "", token.GetSubKind(); expected != actual {
+		return trace.BadParameter("expected sub_kind %v, got %q", expected, actual)
+	}
+	if name := token.GetMetadata().GetName(); name == "" {
+		return trace.BadParameter("missing name")
+	}
+
+	spec := token.GetSpec()
+	if spec == nil {
+		return trace.BadParameter("spec must not be nil")
+	}
+
+	if token.GetScope() == "" {
+		return trace.BadParameter("scoped token must have a scope assigned")
+	}
+
+	if err := scopes.StrongValidate(token.GetScope()); err != nil {
+		return trace.Wrap(err, "validating scoped token resource scope")
+	}
+
+	if spec.AssignedScope != "" {
+		if err := scopes.StrongValidate(spec.AssignedScope); err != nil {
+			return trace.Wrap(err, "validating scoped token assigned scope")
+		}
+		if !scopes.ResourceScope(spec.AssignedScope).IsSubjectToPolicyScope(token.GetScope()) {
+			return trace.BadParameter("scoped token assigned scope must be descendant of its resource scope")
+		}
+	}
+
+	if len(spec.Roles) == 0 {
+		return trace.BadParameter("scoped token must have at least one role")
+	}
+
+	roles, err := types.NewTeleportRoles(spec.Roles)
+	if err != nil {
+		return trace.Wrap(err, "validating scoped token roles")
+	}
+
+	hasBotRole := roles.Include(types.RoleBot)
+	if hasBotRole && spec.BotName == "" {
+		return trace.BadParameter("scoped token with role %q must set bot_name", types.RoleBot)
+	}
+
+	if spec.BotName != "" && !hasBotRole {
+		return trace.BadParameter("can only set bot_name on scoped token with role %q", types.RoleBot)
+	}
+
+	return nil
```

Comment:
Scoped tokens currently support basically no features, but will support the `Node` system role soon.  The `StrongValidate` function should reject bot name and all system roles except `Node`.  We should not be allowing creation of tokens for features that aren't landed or at least in late stage development.